Implement types.Any variadic function handling

mw5h opened the pull request:
#### builtins: Change pg_column_size from Immutable to Stable

pg_column_size() is Stable in Postgres, not Immutable.

Epic: none
Release note: none

#### types: reintroduce the 'any' type

In a previous commit, we converted all usages of the old types.Any to
types.AnyElement, which reflects our legacy use of type.Any. This patch
reintroduces types.Any, which is similar to types.AnyElement, but does
not require all instances of types.Any in a function's argument list to
have the same type (i.e. it's a true wildcard).

Future patches will evaluate whether individual uses of types.AnyElement
should really be using types.Any. For reference, types.Any should be
used whenever the type of a expression is truly unconstrained.

Informs: #136295
Release note: None

#### tree: handle builtin overloads with types.Any variadic arguments

Previously, variadic argument lists with types.Any variable arguments
were being processed as types.AnyElement. An earlier patch changed these
instances to types.AnyElement, correcting this error. This patch now
takes the reintroduced types.Any (unconstrained wildcard) and makes it
available for builtin functions. Much of the overload selection process
is constraining untyped arguments to make variadic types.AnyElement
work, so the implementation here is an early special case to avoid that
processing.

Informs: #136295
Release note: None

#### builtins: change concat() to take types.Any arguments

Previously, concat() took a variable argument list of types.AnyElement.
This patch changes the argument type to types.Any and updates some
related tests.

Fixes: #136295
Release note (sql change): concat() no longer requires all arguments
have the same type.

#### builtins: change num_nulls()/num_notnulls() to take types.Any arguments

Previously, num_nulls() and num_notnulls() required all arguments to
have the same type. This patch removes that restriction.

Epic: none
Release note (sql change): num_nulls and num_notnulls() no longer
require all arguments to have the same type.

#### builtins: change format() to accept types.Any arguments

Previously, format() required all arguments after the format string to
have the same argument type. This patch removes that restriction.

Release note (sql change): format() no longer requires all post-format
string arguments to have the same type.

#### builtins: change crdb_internal.datums_to_bytes() to accept types.Any

Previously, crdb_internal.datums_to_bytes() required all arguments to
have the same type. This patch removes that restriction to allow
arguments of any type.

Release note: None

#### builtins: change concat_ws() to accept arguments of types.Any type

Previously, concat_ws() required all arguments to be of type string,
this patch changes the requirement such that the first argument (the
delimiter) must be a string but that all follow-on arguments may be of
any type.

Release note (sql change): concat_ws() now accepts arguments of any type
in the second and later positions (the separator must still be a
string).

#### builtins: change json_build_objects/array() to accept types.Any

Previously, json_build_objects() and json_build_array() (and their
binary variants) required that all arguments have the same type. This
patch removes that requirement so that arguments can be of any type.

Epic: none
Release note (sql change): json_build_objects(), jsonb_build_objects(),
json_build_array(), jsonb_build_array() no longer require that all
arguments have the same type.

#### builtins: change pg_column_size() to accepts arguments of types.Any

Previously, pg_column_size() required that all arguments have the same
type. This patch removes that restriction so that every argument can
have a different type.

Epic: none
Release note (sql change): pg_column_size() no longer requires that all
arguments have the same type.

mw5h committed:
types: reintroduce the 'any' type

In a previous commit, we converted all usages of the old types.Any to
types.AnyElement, which reflects our legacy use of type.Any. This patch
reintroduces types.Any, which is similar to types.AnyElement, but does
not require all instances of types.Any in a function's argument list to
have the same type (i.e. it's a true wildcard).

Future patches will evaluate whether individual uses of types.AnyElement
should really be using types.Any. For reference, types.Any should be
used whenever the type of a expression is truly unconstrained.

Informs: #136295
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141193)
<!-- Reviewable:end -->

mw5h committed:
tree: handle builtin overloads with types.Any variadic arguments

Previously, variadic argument lists with types.Any variable arguments
were being processed as types.AnyElement. An earlier patch changed these
instances to types.AnyElement, correcting this error. This patch now
takes the reintroduced types.Any (unconstrained wildcard) and makes it
available for builtin functions. Much of the overload selection process
is constraining untyped arguments to make variadic types.AnyElement
work, so the implementation here is an early special case to avoid that
processing.

Informs: #136295
Release note: None

mw5h committed:
builtins: change concat() to take types.Any arguments

Previously, concat() took a variable argument list of types.AnyElement.
This patch changes the argument type to types.Any and updates some
related tests.

Fixes: #136295
Release note (sql change): concat() no longer requires all arguments
have the same type.

mw5h committed:
builtins: Change pg_column_size from Immutable to Stable

pg_column_size() is Stable in Postgres, not Immutable.

Epic: none
Release note: pg_column_size() is now regarded as Stable, to match
Postgres. As a result, it will no longer be allowed in computed column
expressions or partial index predicate expressions.

mw5h committed:
builtins: change num_nulls()/num_notnulls() to take types.Any arguments

Previously, num_nulls() and num_notnulls() required all arguments to
have the same type. This patch removes that restriction.

Epic: none
Release note (sql change): num_nulls and num_notnulls() no longer
require all arguments to have the same type.

mw5h committed:
builtins: change format() to accept types.Any arguments

Previously, format() required all arguments after the format string to
have the same argument type. This patch removes that restriction.

Release note (sql change): format() no longer requires all post-format
string arguments to have the same type.

mw5h committed:
builtins: change crdb_internal.datums_to_bytes() to accept types.Any

Previously, crdb_internal.datums_to_bytes() required all arguments to
have the same type. This patch removes that restriction to allow
arguments of any type.

Release note: None

mw5h committed:
builtins: change concat_ws() to accept arguments of types.Any type

Previously, concat_ws() required all arguments to be of type string,
this patch changes the requirement such that the first argument (the
delimiter) must be a string but that all follow-on arguments may be of
any type.

Release note (sql change): concat_ws() now accepts arguments of any type
in the second and later positions (the separator must still be a
string).

mw5h committed:
builtins: change json_build_objects/array() to accept types.Any

Previously, json_build_objects() and json_build_array() (and their
binary variants) required that all arguments have the same type. This
patch removes that requirement so that arguments can be of any type.

Epic: none
Release note (sql change): json_build_objects(), jsonb_build_objects(),
json_build_array(), jsonb_build_array() no longer require that all
arguments have the same type.

mw5h committed:
builtins: change pg_column_size() to accepts arguments of types.Any

Previously, pg_column_size() required that all arguments have the same
type. This patch removes that restriction so that every argument can
have a different type.

Epic: none
Release note (sql change): pg_column_size() no longer requires that all
arguments have the same type.

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great work and thanks for the organized commits! I left a few minor suggestions.

> Reviewed 1 of 2 files at r1, 8 of 8 files at r3, 8 of 8 files at r4, 1 of 1 files at r5, 6 of 6 files at r6, 4 of 4 files at r7, 4 of 4 files at r8, 2 of 2 files at r9, 4 of 4 files at r10, 4 of 4 files at r11, 3 of 3 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141193#-:-OLjJOxrAzRq50_Ruk9G:b-nprjtu)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2)

___
*[`-- commits` line 7 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141193#-OLjDBf4C9N2xqHvWZtg:-OLjDBf4C9N2xqHvWZth:bmhf0l9):*
nit: Add a release note mentioning the change and that `pg_column_size()` will no longer be allowed in computed column expressions, partial index predicate expressions, etc., as a result of this change.
___
*[`pkg/sql/sem/tree/overload.go` line 934 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/141193#-OLjGULh6A7H_Y8Gyycl:-OLjGULi5Nz5NPEHGe81:b4d3qhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6095199c28199a7a15d2f31cefff4966b398a454/pkg/sql/sem/tree/overload.go#L934)):*
> ```Go
> 			}
> 			for j := range s.exprs {
> 				// We don't have a desired type for the variadic arguments.
> ```

nit: Not sure I understand the comment. VariadicType may have some fixed types preceding the variadic types.Any.
___
*[`pkg/sql/sem/builtins/fixed_oids.go` line 2653 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/141193#-OLjFWxfAY74Z4zcy18L:-OLjFWxgAjb3akzS7Qb4:b-ex5h1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c140a50539311f4cf25fb0b82b96ad2083ab8fa/pkg/sql/sem/builtins/fixed_oids.go#L2653)):*
> ```Go
> 	2688: `crdb_internal.backup_compaction(backup_stmt: string, start_time: decimal, end_time: decimal) -> int`,
> 	2689: `any_send(any: any) -> bytes`,
> 	2690: `any_recv(input: anyelement) -> any`,
> ```

PG does not have `any_send` nor `any_recv`, so we shouldn't either.

It does have `any_out` and `any_in`, but they always error, AFAICT.
___
*[`pkg/sql/logictest/testdata/logic_test/builtin_function` line 3067 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/141193#-OLjHJwZCKFORuWTBhM5:-OLjHJwZCKFORuWTBhM6:b-tyxzge) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ef657c4a666b12dab820ecb86b64f6602071138/pkg/sql/logictest/testdata/logic_test/builtin_function#L3067)):*
> ```
> 
> statement ok
> PREPARE nn_stmt AS SELECT num_nulls(42, $1::STRING, a, b), num_nulls(b, $1, 42, a), num_nulls(a, b, $1, 42), num_nulls($1, a, b, 42), num_nonnulls(42, $1, a, b), num_nonnulls(b, $1, 42, a), num_nonnulls(a, b, $1, 42), num_nonnulls($1, a, b, 42) FROM nulls_test
> ```

nit: use multiple lines for this statement
___
*[`pkg/sql/sem/builtins/builtins.go` line 498 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/141193#-OLjIXcN0YrAD1_gRNgw:-OLjIXcN0YrAD1_gRNgx:bwh7sez) ([raw file](https://github.com/cockroachdb/cockroach/blob/753bebaac0e0fb1eb9e4958da3d399bc6f88cdf9/pkg/sql/sem/builtins/builtins.go#L498)):*
> ```Go
> 						prefix = true
> 					}
> 					d.Format(ctx)
> ```

I wonder if we should be casting the datums to strings instead of directly formatting. We should think about a case where the operations yield different results... I thought a TIMESTAMPTZ might, but it looks like the formatting respects the timezone just like a cast: https://github.com/cockroachdb/cockroach/blob/07f53f8bb0795e29e1139a985b9a1e0964f1ae3e/pkg/sql/sem/tree/datum.go#L3061-L3062
___
*[`pkg/sql/types/types.go` line 597 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/141193#-OLjEGqsACUdbGe79Kd-:-OLjEGqsACUdbGe79Kd0:ba50uis) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c140a50539311f4cf25fb0b82b96ad2083ab8fa/pkg/sql/types/types.go#L597)):*
> ```Go
> 
> 	Any = &T{InternalType: InternalType{
> 		Family: AnyFamily, Oid: oid.T_any, Locale: &emptyLocale}}
> ```

Should Any and AnyElement share the same family? Families are our own invention (we should get rid of them), so I'm having trouble wrapping my head around the implications of these types sharing the same family. I think the biggest implication is with `Equivalent`, which returns true for types that have AnyFamily. So maybe this is fine.


<!-- Sent from Reviewable.io -->

mw5h reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141193)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/sem/builtins/builtins.go` line 498 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/141193#-OLjIXcN0YrAD1_gRNgw:-OLv2IvM3O1-E4764kDH:b1vp0fn) ([raw file](https://github.com/cockroachdb/cockroach/blob/753bebaac0e0fb1eb9e4958da3d399bc6f88cdf9/pkg/sql/sem/builtins/builtins.go#L498)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I wonder if we should be casting the datums to strings instead of directly formatting. We should think about a case where the operations yield different results... I thought a TIMESTAMPTZ might, but it looks like the formatting respects the timezone just like a cast: https://github.com/cockroachdb/cockroach/blob/07f53f8bb0795e29e1139a985b9a1e0964f1ae3e/pkg/sql/sem/tree/datum.go#L3061-L3062
</blockquote></details>

I just did it this way because this is what I did for CONCAT().
___
*[`pkg/sql/sem/builtins/fixed_oids.go` line 2653 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/141193#-OLjFWxfAY74Z4zcy18L:-OLv9uyX8DSEWOWPligU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c140a50539311f4cf25fb0b82b96ad2083ab8fa/pkg/sql/sem/builtins/fixed_oids.go#L2653)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

PG does not have `any_send` nor `any_recv`, so we shouldn't either.

It does have `any_out` and `any_in`, but they always error, AFAICT.
</blockquote></details>

Done.
___
*[`pkg/sql/sem/tree/overload.go` line 934 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/141193#-OLjGULh6A7H_Y8Gyycl:-OLvAKxWD1W0dORxDW3I:bu25kpp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6095199c28199a7a15d2f31cefff4966b398a454/pkg/sql/sem/tree/overload.go#L934)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Not sure I understand the comment. VariadicType may have some fixed types preceding the variadic types.Any.
</blockquote></details>

That was stale. There was a time in the lifespan of this patch where that was hardwired to T_any
___
*[`pkg/sql/types/types.go` line 597 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/141193#-OLjEGqsACUdbGe79Kd-:-OLvASR1CI-_ffA8arbY:b-zgvg53) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c140a50539311f4cf25fb0b82b96ad2083ab8fa/pkg/sql/types/types.go#L597)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Should Any and AnyElement share the same family? Families are our own invention (we should get rid of them), so I'm having trouble wrapping my head around the implications of these types sharing the same family. I think the biggest implication is with `Equivalent`, which returns true for types that have AnyFamily. So maybe this is fine.
</blockquote></details>

For now? They seem at least vaguely related. I'd be amenable to moving them (or just replacing families altogether), but I'm not sure what's better or if giving them their own family would be an improvement.


<!-- Sent from Reviewable.io -->

mw5h commented:
bors r+
